pkg/token/filecache.go

```diff
@@ -0,0 +1,277 @@
+package token
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/gofrs/flock"
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+)
+
+// A mockable filesystem interface
+var f filesystem = osFS{}
+
+type filesystem interface {
+	Stat(filename string) (os.FileInfo, error)
+	ReadFile(filename string) ([]byte, error)
+	WriteFile(filename string, data []byte, perm os.FileMode) error
+}
+
+// default os based implementation
+type osFS struct {}
+
+func (osFS) Stat(filename string) (os.FileInfo, error) {
+	return os.Stat(filename)
+}
+
+func (osFS) ReadFile(filename string) ([]byte, error) {
+	return ioutil.ReadFile(filename)
+}
+
+func (osFS) WriteFile(filename string, data []byte, perm os.FileMode) error {
+	return ioutil.WriteFile(filename, data, perm)
+}
+
+// A mockable environment interface
+var e environment = osEnv{}
+
+type environment interface {
+	Getenv(key string) string
+	LookupEnv(key string) (string, bool)
+}
+
+// default os based implementation
+type osEnv struct {}
+
+func (osEnv) Getenv(key string) string {
+	return os.Getenv(key)
+}
+
+func (osEnv) LookupEnv(key string) (string, bool) {
+	return os.LookupEnv(key)
+}
+
+// A mockable flock interface
+type filelock interface {
+	Unlock() error
+	TryLockContext(ctx context.Context, retryDelay time.Duration) (bool, error)
+	TryRLockContext(ctx context.Context, retryDelay time.Duration) (bool, error)
+}
+
+var newFlock = func(filename string) filelock {
+	return flock.New(filename)
+}
+
+// cacheFile is a map of clusterID/roleARNs to cached tokens
+type cacheFile struct {
+	// a map of clusterIDs/profiles/roleARNs to cachedTokens
+	ClusterMap map[string]map[string]map[string]cacheToken `yaml:"clusters"`
```

Comment:
Let's see.. profile is included in the key because the same role could have multiple entries in your ~/.aws/profile with e.g. different MFA settings?  I think it makes sense just wondering if there is a cleaner representation.

Reply:
I definitely have different profiles for the same cluster, with different IAM roles associated with them.  The aws-iam-authenticator *also* allows profiles to optionally assume different roles via the -r option, if they're able.  I don't use that, in my config, tho.